File: app/src/main/java/social/entourage/android/carousel/CarouselFragment.java
package social.entourage.android.carousel;


import android.content.res.Resources;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.view.ViewPager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.LinearLayout;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import social.entourage.android.base.EntourageDialogFragment;
import social.entourage.android.R;

import static social.entourage.android.carousel.CarouselPageAdapter.NUM_PAGES;

/**
 * Help carousel
 */
public class CarouselFragment extends EntourageDialogFragment {

    // ----------------------------------
    // CONSTANTS
    // ----------------------------------

    public static final String TAG = "social.entourage.android.carousel";

    // ----------------------------------
    // ATTRIBUTES
    // ----------------------------------

    @BindView(R.id.carousel_close_button)
    Button closeButton;

    @BindView(R.id.carousel_view)
    ViewPager pagerView;

    @BindView(R.id.carousel_indicator_layout)
    LinearLayout indicatorLayout;

    List<ImageView> dots;

    // ----------------------------------
    // LIFECYCLE
    // ----------------------------------

    public CarouselFragment() {
        // Required empty public constructor
    }


    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);
        // Inflate the layout for this fragment
        View v = inflater.inflate(R.layout.fragment_carousel, container, false);

        ButterKnife.bind(this, v);

        return v;
    }

    @Override
    public void onViewCreated(final View view, @Nullable final Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        initializeCarousel();
    }

    private void initializeCarousel() {
        CarouselPageAdapter pageAdapter = new CarouselPageAdapter(getChildFragmentManager());
        pagerView.setAdapter(pageAdapter);
        pagerView.addOnPageChangeListener(new ViewPager.OnPageChangeListener() {
            @Override
            public void onPageScrolled(final int position, final float positionOffset, final int positionOffsetPixels) {

            }

            @Override
            public void onPageSelected(final int position) {
                selectDot(position);
            }

            @Override
            public void onPageScrollStateChanged(final int state) {

            }
        });

        addDots();
    }

    private void addDots() {
        dots = new ArrayList<>();

        ImageView dot1 = (ImageView)indicatorLayout.findViewById(R.id.carousel_b1);
        dots.add(dot1);

        ImageView dot2 = (ImageView)indicatorLayout.findViewById(R.id.carousel_b2);
        dots.add(dot2);

        ImageView dot3 = (ImageView)indicatorLayout.findViewById(R.id.carousel_b3);
        dots.add(dot3);

        ImageView dot4 = (ImageView)indicatorLayout.findViewById(R.id.carousel_b4);
        dots.add(dot4);
    }

    public void selectDot(int idx) {
        Resources res = getResources();
        for(int i = 0; i < NUM_PAGES; i++) {
            int drawableId = (i==idx)?(R.drawable.carousel_bullet_filled):(R.drawable.carousel_bullet_empty);
            Drawable drawable = res.getDrawable(drawableId);
            dots.get(i).setImageDrawable(drawable);
        }
    }

    // ----------------------------------
    // BUTTONS HANDLING
    // ----------------------------------

    @OnClick(R.id.carousel_close_button)
    public void onCloseClicked() {
        dismiss();
    }

}

File: app/src/main/java/social/entourage/android/about/carousel/CarouselFragment.kt
package social.entourage.android.about.carousel

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import androidx.appcompat.content.res.AppCompatResources
import androidx.viewpager.widget.ViewPager.OnPageChangeListener
import social.entourage.android.R
import social.entourage.android.base.EntourageDialogFragment
import java.util.*
import kotlinx.android.synthetic.main.fragment_carousel.*
import kotlinx.android.synthetic.main.fragment_carousel.view.*

/**
 * Help carousel
 */
class CarouselFragment  : EntourageDialogFragment() {
    // ----------------------------------
    // ATTRIBUTES
    // ----------------------------------
    private var dots: ArrayList<ImageView> =  ArrayList()

    // ----------------------------------
    // LIFECYCLE
    // ----------------------------------
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        super.onCreateView(inflater, container, savedInstanceState)
        // Inflate the layout for this fragment
        return inflater.inflate(R.layout.fragment_carousel, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initializeCarousel()
    }

    private fun initializeCarousel() {
        carousel_view?.adapter = CarouselPageAdapter(childFragmentManager)
        carousel_view?.addOnPageChangeListener(object : OnPageChangeListener {
            override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {}
            override fun onPageSelected(position: Int) {
                selectDot(position)
            }
            override fun onPageScrollStateChanged(state: Int) {}
        })
        addDots()
        carousel_close_button?.setOnClickListener {dismiss()}
    }

    private fun addDots() {
        carousel_indicator_layout?.let {
            dots.add(it.carousel_b1)
            dots.add(it.carousel_b2)
            dots.add(it.carousel_b3)
            dots.add(it.carousel_b4)   
        }
    }

    fun selectDot(idx: Int) {
        for (i in 0 until CarouselPageAdapter.NUM_PAGES) {
            val drawableId = if (i == idx) R.drawable.carousel_bullet_filled else R.drawable.carousel_bullet_empty
            dots[i].setImageDrawable(AppCompatResources.getDrawable(requireContext(), drawableId))
        }
    }

    companion object {
        // ----------------------------------
        // CONSTANTS
        // ----------------------------------
        const val TAG = "social.entourage.android.about.carousel"
    }
}